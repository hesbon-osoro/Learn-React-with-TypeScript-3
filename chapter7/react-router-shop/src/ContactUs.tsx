import React from 'react';

export interface IProps {
	name: string;
	onNameChange: (name: string) => void;
	email: string;
	onEmailChange: (email: string) => void;
	reason: string;
	onReasonChange: (reason: string) => void;
	notes: string;
	onNotesChange: (notes: string) => void;
}

const ContactUs: React.FC<IProps> = props => {
	const handleNameChange = (e: React.ChangeEvent<HTMLInputElement>) => {
		props.onNameChange(e.currentTarget.value);
	};
	const handleEmailChange = (e: React.ChangeEvent<HTMLInputElement>) => {
		props.onEmailChange(e.currentTarget.value);
	};
	const handleReasonChange = (e: React.ChangeEvent<HTMLSelectElement>) => {
		props.onReasonChange(e.currentTarget.value);
	};
	const handleNotesChange = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
		props.onNotesChange(e.currentTarget.value);
	};

	const handleSubmit = (e: any) => {
		e.preventDefault();
	};
	return (
		<form className="form" noValidate={true} onSubmit={handleSubmit}>
			<div className="form-group">
				<label htmlFor="name">Your name</label>
				<input
					type="text"
					id="name"
					value={props.name}
					onChange={handleNameChange}
				/>
			</div>
			<div className="form-group">
				<label htmlFor="email">Your email address</label>
				<input
					type="email"
					id="email"
					value={props.email}
					onChange={handleEmailChange}
				/>
			</div>
			<div className="form-group">
				<label htmlFor="reason">Reason you need to contact us</label>
				<select id="reason" value={props.reason} onChange={handleReasonChange}>
					<option value="Marketing">Marketing</option>
					<option value="Support" selected>
						Support
					</option>
					<option value="Feedback">Feedback</option>
					<option value="Jobs">Jobs</option>
					<option value="Other">Other</option>
				</select>
			</div>
			<div className="form-group">
				<label htmlFor="notes">Additional notes</label>
				<textarea id="notes" value={props.notes} onChange={handleNotesChange} />
			</div>
		</form>
	);
};

export default ContactUs;
