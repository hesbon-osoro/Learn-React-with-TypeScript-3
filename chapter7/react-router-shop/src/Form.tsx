import * as React from 'react';

export interface IValues {
	[key: string]: any;
}

export interface IFormProps {
	defaultValues: IValues;
}

export interface IState {
	values: IValues;
}

export interface IFieldProps {
	name: string;
	label: string;
	type?: 'Text' | 'Email' | 'Select' | 'TextArea';
	options?: string[];
}

export class Form extends React.Component<IFormProps, IState> {
	constructor(props: IFormProps) {
		super(props);
		this.state = {
			values: props.defaultValues,
		};
	}
	public static Field: React.FC<IFieldProps> = props => {
		return <div></div>;
	};
	public render() {
		return (
			<form className="form" noValidate={true}>
				{this.props.children}
			</form>
		);
	}
}

Form.Field.defaultProps = {
	type: 'Text',
};
