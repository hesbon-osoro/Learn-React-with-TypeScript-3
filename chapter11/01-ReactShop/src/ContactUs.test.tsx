import React from 'react';
import ReactDOM from 'react-dom';
import { Simulate } from 'react-dom/test-utils';
import ContactUs from './ContactUs';
import { ISubmitResult } from './Form';
import { render, cleanup, fireEvent } from '@testing-library/react';

afterEach(cleanup);
describe('ContactUs', () => {
	test('When submit without filling in fields should display errors', () => {
		const handleSubmit = jest.fn();

		const { getAllByText, getByText } = render(
			<ContactUs onSubmit={handleSubmit} />
		);

		const submitButton = getByText('Submit');
		fireEvent.click(submitButton);

		const errorSpans = getAllByText('This must be populated');
		expect(errorSpans.length).toBe(2);

		expect(handleSubmit).not.toBeCalled();
	});
	test('When submit with filling in fields should display okay', () => {
		const handleSubmit = jest.fn();
		const { container, getByText, getByLabelText } = render(
			<ContactUs onSubmit={handleSubmit} />
		);
		const nameField: HTMLInputElement = getByLabelText(
			'Your name'
		) as HTMLInputElement;
		expect(nameField).not.toBeNull();
		fireEvent.change(nameField, { target: { value: 'Hesbon' } });
		const emailField: HTMLInputElement = getByLabelText(
			'Your email address'
		) as HTMLInputElement;
		expect(emailField).not.toBeNull();
		fireEvent.change(emailField, {
			target: { value: 'hesbonosoro1@gmail.com' },
		});
		const submitButton = getByText('Submit');
		fireEvent.click(submitButton);

		const errorsDiv = container.querySelector('[data-testid="formErrors"]');
		expect(errorsDiv).toBeNull();

		expect(handleSubmit).toBeCalledTimes(1);
		expect(handleSubmit).toBeCalledWith({
			name: 'Wazimu',
			email: 'hesbonosoro1@gmail.com',
			reason: 'Support',
			notes: '',
		});
	});
	test('Renders okay', () => {
		const handleSubmit = async (): Promise<ISubmitResult> => {
			return {
				success: true,
			};
		};
		const { container } = render(<ContactUs onSubmit={handleSubmit} />);
		expect(container).toMatchSnapshot();
	});
});
// describe('ContactUs', () => {
// 	test('When submit without filling in fields should display errors', () => {
// 		const handleSubmit = async (): Promise<ISubmitResult> => {
// 			return {
// 				success: true,
// 			};
// 		};
// 		const container = document.createElement('div');
// 		ReactDOM.render(<ContactUs onSubmit={handleSubmit} />, container);
// 		const form = container.querySelector('form');
// 		expect(form).not.toBeNull();
// 		Simulate.submit(form!);
// 		const errorSpans = container.querySelectorAll('.form-error');
// 		expect(errorSpans.length).toBe(3);
// 		ReactDOM.unmountComponentAtNode(container);
// 	});
// });
