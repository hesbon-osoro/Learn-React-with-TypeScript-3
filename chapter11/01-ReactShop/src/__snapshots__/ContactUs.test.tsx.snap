// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactUs Renders okay 1`] = `
<div>
  <form
    class="form"
    novalidate=""
  >
    <div>
       
      <div
        class="form-group"
      >
        <label
          for="name"
        >
          Your name
        </label>
        <input
          id="name"
          type="text"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          for="email"
        >
          Your email address
        </label>
        <input
          id="email"
          type="email"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          for="reason"
        >
          Reason you need to contact us
        </label>
        <select>
          <option
            value="Marketing"
          >
            Marketing
          </option>
          <option
            value="Support"
          >
            Support
          </option>
          <option
            value="Feedback"
          >
            Feedback
          </option>
          <option
            value="Jobs"
          >
            Jobs
          </option>
          <option
            value="Other"
          >
            Other
          </option>
        </select>
      </div>
      <div
        class="form-group"
      >
        <label
          for="notes"
        >
          Additional notes
        </label>
        <textarea
          id="notes"
        />
      </div>
    </div>
    <div
      class="form-group"
    >
      <button
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</div>
`;
